src/java/htsjdk/samtools/sra/SRALazyRecord.java

```diff
@@ -0,0 +1,941 @@
+        if (!initializedFields.contains(LazyField.QUALS)) {
+            initializedFields.add(LazyField.QUALS);
+        }
+        super.setBaseQualities(value);
+    }
+
+    @Override
+    public int getMateAlignmentStart() {
+        if (!initializedFields.contains(LazyField.MATE_ALIGNMENT_START)) {
+            setMateAlignmentStart(getMateAlignmentStartImpl());
+        }
+        return super.getMateAlignmentStart();
+    }
+
+    @Override
+    public void setMateAlignmentStart(final int value) {
+        if (!initializedFields.contains(LazyField.MATE_ALIGNMENT_START)) {
+            initializedFields.add(LazyField.MATE_ALIGNMENT_START);
+        }
+        super.setMateAlignmentStart(value);
+    }
+
+    @Override
+    public String getMateReferenceName() {
+        if (!initializedFields.contains(LazyField.MATE_REFERENCE_NAME)) {
+            setMateReferenceName(getMateReferenceNameImpl());
+        }
+        return super.getMateReferenceName();
+    }
+
+    @Override
+    public void setMateReferenceName(final String value) {
+        if (!initializedFields.contains(LazyField.MATE_REFERENCE_NAME)) {
+            initializedFields.add(LazyField.MATE_REFERENCE_NAME);
+        }
+        super.setMateReferenceName(value);
+    }
+
+    @Override
+    public Integer getMateReferenceIndex() {
+        if (!initializedFields.contains(LazyField.MATE_REFERENCE_NAME)) {
+            setMateReferenceName(getMateReferenceNameImpl());
+        }
+        return super.getMateReferenceIndex();
+    }
+
+    @Override
+    public void setMateReferenceIndex(final int value) {
+        if (!initializedFields.contains(LazyField.MATE_REFERENCE_NAME)) {
+            initializedFields.add(LazyField.MATE_REFERENCE_NAME);
+        }
+        super.setMateReferenceIndex(value);
+    }
+
+    @Override
+    public int getInferredInsertSize() {
+        if (!initializedFields.contains(LazyField.INFERRED_INSERT_SIZE)) {
+            setInferredInsertSize(getInferredInsertSizeImpl());
+        }
+        return super.getInferredInsertSize();
+    }
+
+    @Override
+    public void setInferredInsertSize(final int value) {
+        if (!initializedFields.contains(LazyField.INFERRED_INSERT_SIZE)) {
+            initializedFields.add(LazyField.INFERRED_INSERT_SIZE);
+        }
+        super.setInferredInsertSize(value);
+    }
+
+    // ===== flags =====
+
+    @Override
+    public int getFlags() {
+        for (LazyFlag flag : LazyFlag.values()) {
+            if (initializedFlags.contains(flag)) {
+                continue;
+            }
+
+            FlagGetter flagGetter = lazyFlagsGetters.get(flag);
+            if (flagGetter == null) {
+                throw new RuntimeException("No flag " + flag + " getter specified");
+            }
+
+            try {
+                if (flagGetter.canCallOnNotPaired || getReadPairedFlag()) {
+                    this.getClass().getMethod(flagGetter.getterName).invoke(this);
```

Comment:
Invoking methods via reflection like this is much slower than invoking them directly. If htsjdk supported Java 8 we could use method references, but since it doesn't yet what you can do instead is have an abstract `getFlag()` operation in your `LazyFlag()` enum, and an implementation for each enum value that calls into the correct method.